Lib/test/test_zipfile.py

```diff
@@ -2724,16 +2724,71 @@ def test_extract_command(self):
                                 self.assertEqual(f.read(), zf.read(zi))
 
 
+class TestExecutablePrependedZip(unittest.TestCase):
+    """Test our ability to open zip files with an executable prepended."""
+
+    def setUp(self):
+        self.exe_zip = findfile('exe_with_zip', subdir='ziptestdata')
+        self.exe_zip64 = findfile('exe_with_z64', subdir='ziptestdata')
+
+    def _test_zip_works(self, name):
+        # bpo28494 sanity check: ensure is_zipfile works on these.
+        self.assertTrue(zipfile.is_zipfile(name),
+                        f'is_zipfile failed on {name}')
+        # Ensure we can operate on these via ZipFile.
+        with zipfile.ZipFile(name) as zipfp:
+            for n in zipfp.namelist():
+                data = zipfp.read(n)
+                self.assertIn(b'FAVORITE_NUMBER', data)
+
+    def test_read_zip_with_exe_prepended(self):
+        self._test_zip_works(self.exe_zip)
+
+    def test_read_zip64_with_exe_prepended(self):
+        self._test_zip_works(self.exe_zip64)
+
+    @unittest.skipUnless(sys.executable, 'sys.executable required.')
+    @unittest.skipUnless(os.access('/bin/bash', os.X_OK),
+                         'Test relies on #!/bin/bash working.')
+    def test_execute_zip2(self):
+        output = subprocess.check_output([self.exe_zip, sys.executable])
+        self.assertIn(b'number in executable: 5', output)
+
+    @unittest.skipUnless(sys.executable, 'sys.executable required.')
+    @unittest.skipUnless(os.access('/bin/bash', os.X_OK),
+                         'Test relies on #!/bin/bash working.')
+    def test_execute_zip64(self):
+        output = subprocess.check_output([self.exe_zip64, sys.executable])
+        self.assertIn(b'number in executable: 5', output)
+
+
 # Poor man's technique to consume a (smallish) iterable.
 consume = tuple
 
 
+# from jaraco.itertools 5.0
+class jaraco:
```

Comment:
This could probably named better. I guess it's just calling iter on the given object and used only once. Can this be replaced by just using iter on the object to be consumed by for loop?

Reply:
I used `jaraco.itertools.Counter` because that's [what it's called in the backport](https://github.com/jaraco/zipp/blob/f73569f7d887c7a0f0ecc51e8d09d42ac272edf1/test_zipp.py#L223). I don't see how it can be replaced by using iter on the object. The idea behind this class is to wrap an iterable and keep a count of how many iterations it's been through. It's true you can do that with a few extra lines in every code that does this, but then you run the risk of off-by-one errors each time you try to do it and requires entangling the iteration with the counting, whereas this approach keeps the two concerns separated in a re-usable implementation. I'd rather re-use this functionality rather than inlining it in yet another place.

It sort-of feels to me like maybe your comment was associated with the wrong line.

Reply:
Sorry, I missed the part where count is used in the test. My part about naming was to just have this as `Counter` as a top level class instead of `jaraco.itertools.Counter` for the test.

Reply:
Okay, good. In that case, let's stick with the namespace for now. Changing it here would require changing it in the backport (or maintaining the variance), and I'd rather move forward with this change. Thanks for the feedback.